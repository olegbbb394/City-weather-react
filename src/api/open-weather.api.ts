import {ICityWeatherItem} from "../types/city-weather.model";

export const getWeatherById = async (name: string): Promise<ICityWeatherItem | null> => {
    const param = {
        "key": "https://api.openweathermap.org/data/2.5/",
        "appid": "1978d5cd2ed3658747a2c546f5ef164e"
    }

    const url = `${param.key}weather?q=${name}&units=metric&APPID=${param.appid}`;

    try {
        const response = await fetch(url);

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`)
        }

        return await response.json();
    } catch (error) {
        return null;
    }
};
