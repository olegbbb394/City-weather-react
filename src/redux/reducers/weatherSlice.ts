import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from "../store/store";
import {ICityWeatherItem} from "../../types/city-weather.model";

export interface IWeatherState {
    weatherDataArr: ICityWeatherItem[];
}

const initialState: IWeatherState = {
    weatherDataArr: [],
};

const weatherSlice = createSlice({
    name: 'weather',
    initialState,
    reducers: {
        setWeatherDataArr: (state: IWeatherState, action: PayloadAction<ICityWeatherItem[]>) => {
            state.weatherDataArr = action.payload;
        },
    },
});

export const {setWeatherDataArr} = weatherSlice.actions;
export const weatherReducer = weatherSlice.reducer;
export const selectWeatherDataArr = (state: RootState) => state.weather.weatherDataArr;