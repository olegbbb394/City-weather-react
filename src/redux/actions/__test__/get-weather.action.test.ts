import {ICityWeatherItem} from "../../../types/city-weather.model";
import {setWeatherDataArr} from "../../reducers/weatherSlice";
import {getWeatherById} from "../../../api/open-weather.api";
import {getWeatherAction} from "../get-weather.action";

jest.mock("../../../api/open-weather.api");
const getWeatherByIdMock = getWeatherById as unknown as jest.Mock<typeof getWeatherById>;

const data: ICityWeatherItem | null = {
    id: 703448,
    main: {
        pressure: 1000,
        temp: 35,
    },
    name: 'Kyiv',
    weather: [{
        description: 'good',
        icon: 'o2d',
        id: 2,
        main: 'good',
    }],
    wind: {
        speed: 20,
    },
};

describe('getWeatherAction', () => {
    it('should dispatch setWeatherDataArr with weather data', async () => {
        const thunk = getWeatherAction('Kyiv');
        const dispatchMock = jest.fn();
        const getStateMock = jest.fn();

        getWeatherByIdMock.mockResolvedValueOnce(data as never);
        await thunk(dispatchMock, getStateMock, {});

        expect(dispatchMock).toBeCalledTimes(1);
        expect(dispatchMock).toHaveBeenNthCalledWith(1, setWeatherDataArr([data]));
        expect(dispatchMock).toMatchSnapshot();
    });

    it('should not dispatch setWeatherDataArr when getWeatherById returns null', async () => {
        const thunk = getWeatherAction('Kyiv');
        const dispatchMock = jest.fn();
        const getStateMock = jest.fn();

        getWeatherByIdMock.mockResolvedValueOnce(null!);
        await thunk(dispatchMock, getStateMock, {});

        expect(dispatchMock).not.toBeCalled();
        expect(dispatchMock).toMatchSnapshot();
    });
});