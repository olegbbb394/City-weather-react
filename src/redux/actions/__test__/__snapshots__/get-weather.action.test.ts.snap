// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWeatherAction should dispatch setWeatherDataArr with weather data 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Array [
          Object {
            "id": 703448,
            "main": Object {
              "pressure": 1000,
              "temp": 35,
            },
            "name": "Kyiv",
            "weather": Array [
              Object {
                "description": "good",
                "icon": "o2d",
                "id": 2,
                "main": "good",
              },
            ],
            "wind": Object {
              "speed": 20,
            },
          },
        ],
        "type": "weather/setWeatherDataArr",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`getWeatherAction should not dispatch setWeatherDataArr when getWeatherById returns null 1`] = `[MockFunction]`;
